src/lib/es2023.collection.d.ts

```diff
@@ -0,0 +1,16 @@
+interface WeakKeyTypesStore {
+    symbol: symbol;
+}
+
+interface WeakSet<T extends WeakKeyTypes> {}
+
+interface WeakSetConstructor {
+    new <T extends WeakKeyTypes = WeakKeyTypes>(values?: readonly T[] | null): WeakSet<T>;
+}
+
+interface WeakMap<K extends WeakKeyTypes, V> {}
+
+interface WeakMapConstructor {
+    new <K extends WeakKeyTypes = WeakKeyTypes, V = any>(entries?: readonly [K, V][] | null): WeakMap<K, V>;
+    readonly prototype: WeakMap<WeakKeyTypes, any>;
+}
```

Comment:
I'm wondering. Instead of adding overloads, maybe we should modify the original definitions? https://github.com/bloomberg/TypeScript/blob/18ec290e15e6de05fe09872d5de8bd8d11c3ede2/src/lib/es2015.collection.d.ts#L73

```suggestion
```

Reply:
I am not sure what the best approach here would be, I thought it generally was a bit frowned upon to make changes in past es versions?

Reply:
Yeah, in terms of what JavaScript features they are say are available they are fixed. But from a 'types' perspective, the types can still be 'improved' on as TypeScript itself evolves.

Thanks for making the change, though now you've done that I notice that you had to update the various redeclarations of WeakMap and WeatSet across [collection](https://github.com/microsoft/TypeScript/blob/13c374a868c926f6a907666a5599992c1351b773/src/lib/es2015.collection.d.ts#L45), [iterable](https://github.com/microsoft/TypeScript/blob/13c374a868c926f6a907666a5599992c1351b773/src/lib/es2015.iterable.d.ts#L144) and [symbol.wellknow](https://github.com/microsoft/TypeScript/blob/13c374a868c926f6a907666a5599992c1351b773/src/lib/es2015.symbol.wellknown.d.ts#L122). With `WeakKey` only defined in [es2015.collection.d.ts](https://github.com/bloomberg/TypeScript/blob/0809ac22893653d3c5494e89bae90351d878b6b9/src/lib/es2015.collection.d.ts#L49) this means that `es2015.iterable.d.ts` and `es2015.symbol.wellknown.d.ts` will have any error if `es2015.collection.d.ts` is also not being included because `WeakKey` won't have been defined.

Looking for inspiration on how might be best to solve this. I noticed that if I only loaded `es2015.iterable.d.ts` into a project I got errors [here](https://github.com/microsoft/TypeScript/blob/13c374a868c926f6a907666a5599992c1351b773/src/lib/es2015.iterable.d.ts#L207) because the type `Awaited` did not exist, as it is defined in [es5.d.ts](https://github.com/microsoft/TypeScript/blob/13c374a868c926f6a907666a5599992c1351b773/src/lib/es5.d.ts#L1524).

So we could copy how `Awaited` is setup, and move:

```ts
interface WeakKeyTypes {
    object: object;
}

type WeakKey = WeakKeyTypes[keyof WeakKeyTypes];
```

into `es5.d.ts`, ready to be used across the various files in es2015. Alternatively we could make `es2015.iterable.d.ts` and `es2015.symbol.wellknown.d.ts` _import_ on `es2015.collection.d.ts` by adding `/// <reference lib="es2015.collection" />` to the top of them. Though this might cause more issues if anyone is specifically including `es2015.iterable` because they don't want to pull in all of `es2015.collection` too.
We can then also remove the repeated definitions as they match the ones already in `es2015`, as per [the suggestion here](https://github.com/bloomberg/TypeScript/pull/76/files/18ec290e15e6de05fe09872d5de8bd8d11c3ede2#r1149164574)